rust-runtime/aws-smithy-runtime/src/client/orchestrator/auth.rs

```diff
@@ -20,18 +20,78 @@ use std::error::Error as StdError;
 use std::fmt;
 use tracing::trace;
 
+#[derive(Debug)]
+struct NoMatchingAuthSchemeError(ExploredList);
+
+impl fmt::Display for NoMatchingAuthSchemeError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let explored = &self.0;
+
+        // Use the information we have about the auth options that were explored to construct
+        // as helpful of an error message as possible.
+        if explored.items().count() == 0 {
+            return f
+                .write_str("no auth options are available. This is a bug. Please file an issue.");
+        }
+        if explored
+            .items()
+            .all(|explored| matches!(explored.result, ExploreResult::NoAuthScheme))
+        {
+            return f
+                .write_str("no auth schemes are registered. This is a bug. Please file an issue.");
+        }
+
+        let mut try_add_identity = false;
+        let mut likely_bug = false;
+        f.write_str("failed to select an auth scheme to sign the request with.")?;
+        for item in explored.items() {
+            write!(
+                f,
+                " \"{}\" wasn't an option because ",
+                item.scheme_id.as_str()
+            )?;
+            f.write_str(match item.result {
+                        ExploreResult::NoAuthScheme => {
+                            likely_bug = true;
+                            "no auth scheme was registered for it."
+                        },
+                        ExploreResult::NoIdentityResolver => {
+                            try_add_identity = true;
+                            "there was no identity resolver for it."
+                        }
+                        ExploreResult::MissingEndpointConfig => {
+                            likely_bug = true;
+                            "it has auth config in its endpoint config, but this scheme wasn't listed in that endpoint config."
```

Comment:
I don't totally know what this means 😓 . What if we link the error message to a discussion on smithy-rs?